libretro-rs/src/lib.rs

```diff
@@ -829,19 +876,23 @@ impl<T: RetroCore> RetroInstance<T> {
   pub fn on_get_memory_data(&mut self, id: libc::c_uint) -> *mut () {
     let mut env = self.environment();
     self.core_mut(|core| {
-      if let Some(data) = core.get_memory_data(&mut env, id) {
-        // TODO: is there a way to maintain lifetimes here?
-        data.as_mut_ptr() as *mut ()
-      } else {
-        std::ptr::null_mut()
-      }
+      u16::try_from(id)
+        .ok()
+        .and_then(|id| RetroMemoryType::try_from(id).ok())
+        .and_then(|id| core.get_memory_data(&mut env, id))
+        .map(|data| data.as_mut_ptr() as *mut ())
+        .unwrap_or(std::ptr::null_mut())
     })
   }
 
   /// Invoked by a `libretro` frontend, with the `retro_get_memory_size` API call.
   pub fn on_get_memory_size(&mut self, id: libc::c_uint) -> libc::size_t {
     let mut env = self.environment();
-    self.core_mut(|core| core.get_memory_size(&mut env, id))
+    u16::try_from(id)
+      .ok()
+      .and_then(|id| RetroMemoryType::try_from(id).ok())
+      .map(|id| self.core_mut(|core| core.get_memory_size(&mut env, id)))
+      .unwrap_or(0)
```

Comment:
Not sure if this helps or not, but I believe you can declare a local function that returns an error type (`Option<T>` or `Result<T, E>`) and use the fancy `?` syntax, then invoke it. In this case it would look like:

```rust
pub fn on_get_memory_size(&mut self, id: libc::c_uint) -> libc::size_t {
  fn do_the_thing() -> Option<libc::size_t> {
    let id = u16::try_from(id).ok()?;
    let mem_type = RetroMemoryType::try_from(id).ok()?;
    let mut env = self.environment();
    self.core_mut(|core| core.get_memory_size(&mut env, mem_type))
  }

  do_the_thing().unwrap_or(0)
}
```

Reply:
Thanks for reminding me about the `?` operator, the fact that it works with `Option` isn't as well-documented as it is for `Result`.  I found a more compact way to rewrite the code but I'll keep this in mind going forward. Let me know what you think of the new version.